### PR TITLE
Header-only refactoring of trajectory_distances_and_speeds

### LINKED ISSUE #563: [FEA] Refactor trajectory APIs to header-only API
Create header-only APIs for the following and refactor existing APIs / tests on top of the header-only API:

- [x] `cuspatial::derive_trajectories` #628 
- [ ] `cuspatial::trajectory_distances_and_speeds`
- [x] `cuspatial::trajectory_bounding_boxes`

See https://github.com/rapidsai/cuspatial/blob/branch-22.08/cpp/include/cuspatial/trajectory.hpp

https://github.com/rapidsai/cuspatial/issues/563#tasklist-block-0ad734e5-e92a-454a-bf8e-72701ce77eb6

### PR DESCRIPTION
**Pull request by harrism**
## Description
Closes #563 

In addition to refactoring to header-only API, this PR changes the implementation to use `thrust_transform` (two input version) instead of `thrust::adjacent_difference`. This simplifies the code a bit and eliminates the need for the `duplicate_first_element_iterator`, which was a bit confusing.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cuspatial/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/include/cuspatial/vec_2d.hpp [hunk 1]
```diff
@@ -53,6 +53,18 @@ std::ostream& operator<<(std::ostream& os, cuspatial::vec_2d<T> const& vec)
   return os << "(" << vec.x << "," << vec.y << ")";
 }
 
+/**
+ * @brief Print function usable in host or device code.
+ */
+template <typename T>
+void CUSPATIAL_HOST_DEVICE print(cuspatial::vec_2d<T> const& vec)
```
Comment (isVoid) [inline cpp/include/cuspatial/vec_2d.hpp:60]: I don't think we really need this utility other than debugging in the library?

File: cpp/include/cuspatial/experimental/detail/trajectory_distances_and_speeds.cuh [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+  // Reduce the intermediate durations and kilometer distances into meter
+  // distances and speeds in meters/second
+  thrust::reduce_by_key(rmm::exec_policy(stream),
+                        ids_first,  // keys
+                        ids_last,
+                        duration_and_distance_tmp,       // values
+                        thrust::discard_iterator(),      // keys_output
+                        duration_distances_and_speed,    // values_output
+                        thrust::equal_to<Id>(),          // binary_pred
+                        [] __device__(auto a, auto b) {  // binary_op
```
Comment (isVoid) [inline cpp/include/cuspatial/experimental/detail/trajectory_distances_and_speeds.cuh:121]: The trailing comments can be removed. Depending on the IDE one uses, the name of the function arguments might be shown as UI.

File: cpp/tests/experimental/trajectory/trajectory_distances_and_speeds_test.cu [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+                     [] __device__(T const& a, T const& b) { return max(abs(a), abs(b)); });
+
+    // We expect the floating point error (in ulps) due to be proportional to the  number of
+    // operations to compute the relevant quantity. For distance, this is computation is
+    // m_per_km * sqrt(dot(vec, vec)), where vec = (p1 - p0).
+    // For speed, there is an additional division. There is also accumulated error in the reductions
+    // and we find k_ulps == 10 reliably results in the expected computation matching the actual
+    // computation for large trajectories with disparate positions and increasing timestamps.
+    // This value and the magnitude of the values involed (e.g. max distance) are used to scale
+    // the machine epsilon to come up with an absolute error tolerance.
```
Comment (isVoid) [inline cpp/tests/experimental/trajectory/trajectory_distances_and_speeds_test.cu:77]: :+1:

File: cpp/tests/trajectory/test_trajectory_distances_and_speeds.cu [hunk 4]
```diff
@@ -14,137 +14,97 @@
...
+    auto id = cudf::column(rmm::device_uvector<cudf::size_type>(size, rmm::cuda_stream_default));
+    auto xs = cudf::column(rmm::device_uvector<float>(size, rmm::cuda_stream_default));
+    auto ys = cudf::column(rmm::device_uvector<double>(size, rmm::cuda_stream_default));
+    auto ts = cudf::column(rmm::device_uvector<cudf::timestamp_ms>(size, rmm::cuda_stream_default));
+    EXPECT_THROW(cuspatial::trajectory_distances_and_speeds(1, id, xs, ys, ts, this->mr()),
+                 cuspatial::logic_error);
+  }
 }
 
-TYPED_TEST(TrajectoryDistanceSpeedTest, ComputeDistanceAndSpeed3Simple)
```
Comment (isVoid) [inline cpp/tests/trajectory/test_trajectory_distances_and_speeds.cu:112]: There are a few hand derived tests in original test sets. Should we try to reintroduce them in the tests for header only API?

File: cpp/include/cuspatial/experimental/trajectory_distances_and_speeds.cuh [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+ * @tparam OutputIt Iterator over output (distance, speed) pairs. Must meet the requirements of
+ * [LegacyRandomAccessIterator][LinkLRAI] and be device-writeable.
+ * @tparam IndexT The type of the object IDs.
+ *
+ * @param num_trajectories number of trajectories (unique object ids)
+ * @param ids_first beginning of the range of input object ids
+ * @param ids_last end of the range of input object ids
+ * @param points_first beginning of the range of input point (x,y) coordinates
+ * @param timestamps_first beginning of the range of input timestamps
+ * @param distances_and_speeds beginning of the range of output (distance, speed) pairs
```
Comment (isVoid) [inline cpp/include/cuspatial/experimental/trajectory_distances_and_speeds.cuh:49]:
```suggestion
 * @param distances_and_speeds_first beginning of the range of output (distance, speed) pairs
```

File: cpp/include/cuspatial_test/test_util.cuh [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+#include <thrust/host_vector.h>
+
+#include <cstdio>
+
+namespace cuspatial {
+
+namespace test {
+
+template <typename T, typename Vector>
+thrust::host_vector<T> to_host(Vector const& dvec)
```
Comment (isVoid) [inline cpp/include/cuspatial_test/test_util.cuh:36]: Docs for these utility functions?

File: cpp/tests/experimental/trajectory/trajectory_distances_and_speeds_test.cu (hunk 3 above)
Comment (isVoid) [reply cpp/tests/experimental/trajectory/trajectory_distances_and_speeds_test.cu:77]: Should probably move the comment before `max_expected_distance`.

**Review by isVoid — COMMENTED** [review]
Seems pretty solid to me. I have one question regarding reintroducing the hand derived tests.

File: cpp/include/cuspatial/vec_2d.hpp (hunk 1 above)
Comment (harrism) [reply cpp/include/cuspatial/vec_2d.hpp:60]: Is that not a good reason to have it? We have operator<< already, but that didn't work in device code so I added this.

File: cpp/tests/trajectory/test_trajectory_distances_and_speeds.cu (hunk 4 above)
Comment (harrism) [reply cpp/tests/trajectory/test_trajectory_distances_and_speeds.cu:112]: I felt they were redundant to the new tests.

**Comment by isVoid** [comment]
rerun tests

File: cpp/include/cuspatial/vec_2d.hpp (hunk 1 above)
Comment (isVoid) [reply cpp/include/cuspatial/vec_2d.hpp]: Cool, I think it'll come handy when I need it.

**Review by isVoid — APPROVED** [review]
lgtm!

File: cpp/include/cuspatial/vec_2d.hpp (hunk 1 above)
Comment (isVoid) [reply cpp/include/cuspatial/vec_2d.hpp]: Ah, you removed them in the latest commit. I think it's ok either way.

**Comment by harrism** [comment]
@gpucibot merge

**Pull request merged**